File: src/0015_three_sum/three_sum.go
/*
date: 2021-05-22  12:50
author: yqq
descriptions: three sum
*/

package main

import (
	"fmt"
	"sort"
)

func threeSum(nums []int) [][]int {
	result := [][]int{}
	sort.Ints(nums)
	for i := 0; i < len(nums) - 2; i++ {
		if nums[i] > 0 {
			break
		}

		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		a := nums[i]
		left, right := i+1, len(nums)-1
		for left < right {
			b, c := nums[left] , nums[right]
			sum := a + b + c
			if sum < 0 {
				left++
			} else if sum > 0 {
				right--
			} else if sum == 0 {
				result = append(result, []int{nums[i], nums[left], nums[right]})

				// 跳过重复的
				for left < right && nums[left] == b { left++ }
				for left < right && nums[right] == c { right-- }
			}
		}
	}

	return result
}

func test(nums []int) {
	// sort.Ints(nums)
	// fmt.Printf("%v", nums)
	result := threeSum(nums)
	for _, v := range result {
		for _, item := range v {
			fmt.Printf("%d ", item)
		}
		fmt.Printf("\n")
	}
	fmt.Printf("========\n")
}

func main() {

	test([]int{-1, 0, 1, 2, -1, -4})
	test([]int{3, 0, -2, -1, 1, 2})
	test([]int{1, 2, -2, -1})
	test([]int{-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4})
}

File: src/0015_three_sum/three_sum.cc
// Author: yqq
// Date: 2021-05-21 18:10:24
// Description: three nums sum

#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
#include <set>
using namespace std;

class Solution
{
public:
    // vector<vector<int>> threeSum(vector<int>& nums) {
    //     vector<vector<int>> result;
    //     set<vector<int> > resultSet;
    //     std::sort(nums.begin(), nums.end(), [](int a, int b){return a < b;});
    //     for(int i = 0; i < nums.size(); i++)
    //     {
    //         int a = nums[i];
    //         for(int j = i+1; j < nums.size(); j++)
    //         {
    //             int b = nums[j];
    //             for(int k = j+1; k < nums.size(); k++)
    //             {
    //                 int c = nums[k];
    //                 if(0 == nums[i] + nums[j] + nums[k] )
    //                 {
    //                     auto oneResult = vector<int>({a, b, c});
    //                     if(resultSet.find(oneResult) == resultSet.end())
    //                     {
    //                         resultSet.insert(oneResult);
    //                         result.emplace_back(oneResult);
    //                     }
    //                 }
    //             }
    //         }
    //     }

    //     return result;
    // }

    vector<vector<int>> threeSum(vector<int> &nums)
    {
        set<vector<int>> resultSet;
        std::sort(nums.begin(), nums.end(), [](int a, int b) { return a < b; });

        for (int i = 0; i < nums.size(); i++)
        {
            if(nums[i] > 0) continue;

            if (i > 0 && nums[i] == nums[i - 1]) 
            {
                continue;
            }

            int left = i+ 1;
            int right = nums.size() - 1;
            for(;left < right;)
            {
                int sum = nums[i] + nums[left] + nums[right];
                if(sum > 0) 
                {
                    right--;
                }
                else if(sum < 0)
                {
                    left++;
                }
                if (sum == 0)
                {
                    resultSet.insert({nums[left], nums[i], nums[right]});
                    right--; 
                    left++;
                    continue;
                }
            }
        }
        return vector<vector<int>>(resultSet.begin(), resultSet.end());
    }
};

void test(vector<int> &nums)
{
    auto vctResult = Solution().threeSum(nums);
    for (auto v : vctResult)
    {
        for (auto item : v)
        {
            cout << item << " ";
        }
        cout << endl;
    }
    cout << "================" << endl;
}

int main()
{

    vector<int> vctNums({-1, 0, 1, 2, -1, -4});
    test(vctNums);
    // // // -4, -1, -1, 0, 1, 2
    vector<int> vctNums2({3, 0, -2, -1, 1, 2});
    test(vctNums2);

    vector<int> vctNums3({1, 2, -2, -1});
    test(vctNums3);

    vector<int> vctNums4({-1,0,1,2,-1,-4,-2,-3,3,0,4});
    test(vctNums4);

    return 0;
}
